tests/web/test_db_query.py

```diff
@@ -93,8 +108,107 @@ def test_apply_thread_quota(
     for rlimit in rate_limit_params:
         settings.add_rate_limit(rlimit)
     settings.set_resource_quota(resource_quota)
-    clickhouse_query_settings: MutableMapping[str, Any] = {}
+    clickhouse_query_settings: dict[str, Any] = {}
     _apply_thread_quota_to_clickhouse_query_settings(
         settings, clickhouse_query_settings, rate_limit_stats
     )
     assert clickhouse_query_settings == expected_query_settings
+
+
+def test_raw_query_success() -> None:
+    storage = get_storage(StorageKey("errors"))
+    query = ClickhouseQuery(
+        from_clause=Table(
+            storage.get_schema().get_data_source().get_table_name(),  # type: ignore
+            schema=storage.get_schema().get_columns(),
+            final=False,
+        ),
+        selected_columns=[
+            SelectedExpression(
+                "porject_count",
+                parse_clickhouse_function("count(distinct(project_id))"),
+            )
+        ],
+    )
+
+    query_metadata_list: list[ClickhouseQueryMetadata] = []
+    stats: dict[str, Any] = {}
+
+    result = raw_query(
+        clickhouse_query=query,
+        query_settings=HTTPQuerySettings(),
+        attribution_info=AttributionInfo(
+            app_id=AppID(key="key"),
+            tenant_ids={},
+            referrer="something",
+            team=None,
+            feature=None,
+            parent_api=None,
+        ),
+        dataset_name="events",
+        query_metadata_list=query_metadata_list,
+        formatted_query=format_query(query),
+        reader=storage.get_cluster().get_reader(),
+        timer=Timer("foo"),
+        stats=stats,
+        trace_id="trace_id",
+        robust=False,
+    )
+    assert len(query_metadata_list) == 1
+    assert result.extra["stats"] == stats
+    assert result.extra["sql"] is not None
+    assert set(result.result["profile"].keys()) == {  # type: ignore
+        "elapsed",
+        "bytes",
+        "blocks",
+        "rows",
+    }
+
+
+def test_raw_query_fail() -> None:
+    storage = get_storage(StorageKey("errors"))
+    query = ClickhouseQuery(
+        from_clause=Table(
+            storage.get_schema().get_data_source().get_table_name(),  # type: ignore
+            schema=storage.get_schema().get_columns(),
+            final=False,
+        ),
+        selected_columns=[
+            SelectedExpression(
+                "bad_column", parse_clickhouse_function("count(non_existent_column)")
+            )
+        ],
+    )
+
+    query_metadata_list: list[ClickhouseQueryMetadata] = []
```

Comment:
nit `list` -> `MutableSequence`

Reply:
`list` from future annotations > `MutableSequence` imo

Reply:
gonna ignore this nit. it is a list after all